service/history/historyEngine.go

```diff
@@ -2254,20 +2255,93 @@ type updateWorkflowAction struct {
 	createDecision bool
 }
 
-func (e *historyEngineImpl) updateWorkflowExecutionWithAction(
+type workflowLockedContext struct {
+	context workflowExecutionContext
+	release releaseWorkflowExecutionFunc
+	err     error
+}
+
+type updateWorkflowActionFunc func(builder mutableState) (*updateWorkflowAction, error)
+
+func (e *historyEngineImpl) updateWorkflowExecutionWithActionAndContextReload(
 	ctx ctx.Context,
 	domainID string,
 	execution workflow.WorkflowExecution,
-	action func(builder mutableState) (*updateWorkflowAction, error),
+	action updateWorkflowActionFunc,
 ) (retError error) {
 
+	// for execution with runID, just update with action
+	if execution.GetRunId() != "" {
+		workflowContext := e.getWorkflowLockedContext(ctx, domainID, execution)
+		return e.updateWorkflowExecutionWithContextAndAction(workflowContext, action)
```

Comment:
missing a defer here?

Reply:
not really. context lock is released when perform action inside `updateWorkflowExecutionWithContextAndAction`